### PR TITLE
Add `quote` and `escape` attributes to create csv external table

### LINKED ISSUE #8310: Add `quote` & `escape` for create external table
### Is your feature request related to a problem or challenge?

1、Add `quote` & `escape` for create external table
2、In addition, the DELIMITER attribute is only used in csv. Can it also be moved to option?

### Describe the solution you'd like

```sql
CREATE EXTERNAL TABLE custom(
    c1  VARCHAR DEFAULT NULL,
    c2  VARCHAR DEFAULT NULL
)
STORED AS CSV
WITH HEADER ROW
OPTIONS ('delimiter' ',','read.escape' '\', 'read.quote' '~')
LOCATION 'custom.csv';
```
> `read.escape` & `read.quote` from [https://github.com/apache/arrow-datafusion/pull/8251 comment](https://github.com/apache/arrow-datafusion/pull/8251#discussion_r1401004666)

### Describe alternatives you've considered

No

### Additional context

Related: https://github.com/apache/arrow-datafusion/pull/8251

### PR DESCRIPTION
**Pull request by Asura7969**
## Which issue does this PR close?

Closes #8310.

## Rationale for this change

Create csv table in sql mode, support quote & escape

## What changes are included in this PR?

Add `quote` and `escape` attributes to create csv external table
```sql
CREATE EXTERNAL TABLE custom(
    c1  VARCHAR DEFAULT NULL,
    c2  VARCHAR DEFAULT NULL
)
STORED AS CSV
DELIMITER ','
WITH HEADER ROW
OPTIONS ('escape' '\', 'quote' '~')
LOCATION 'custom.csv';
```

## Are these changes tested?

yes, `csv_files.slt`

## Are there any user-facing changes?

Add `quote` & `escape` attributes to **OPTIONS**.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/csv_files.slt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+statement ok
+CREATE EXTERNAL TABLE csv_with_escape (
+c1 VARCHAR,
+c2 VARCHAR
+) STORED AS CSV
+WITH HEADER ROW
+DELIMITER ','
+OPTIONS ('escape' '\"')
+LOCATION '../core/tests/data/escape.csv';
```
Comment (Asura7969) [inline datafusion/sqllogictest/test_files/csv_files.slt:37]: To `CREATE EXTERNAL TABLE` syntax coverage, I add this file

File: datafusion/common/src/file_options/csv_writer.rs [hunk 2]
```diff
@@ -91,6 +91,11 @@ impl TryFrom<(&ConfigOptions, &StatementOptions)> for CsvWriterOptions {
                         )
                     })?)
             },
+                "quote" | "escape" => {
+                    // These two attributes are only available when reading csv files.
```
Comment (alamb) [inline datafusion/common/src/file_options/csv_writer.rs:95]: Yeah, this is an gap in the arrow CSV handling. I filed https://github.com/apache/arrow-rs/issues/5146

```suggestion
                    // https://github.com/apache/arrow-rs/issues/5146
                    // These two attributes are only available when reading csv files.
```

File: datafusion/core/src/datasource/listing_table_factory.rs [hunk 3]
```diff
@@ -67,12 +67,23 @@ impl TableProviderFactory for ListingTableFactory {
...
-                CsvFormat::default()
-                    .with_has_header(cmd.has_header)
-                    .with_delimiter(cmd.delimiter as u8)
-                    .with_file_compression_type(file_compression_type),
-            ),
+            FileType::CSV => {
+                let mut statement_options = StatementOptions::from(&cmd.options);
+                let quote = statement_options
+                    .take_str_option("quote")
+                    .map_or(b'"', |x| x.as_bytes()[0]);
```
Comment (alamb) [inline datafusion/core/src/datasource/listing_table_factory.rs:74]: This effectively overrides the default values in the CSV format.

What would you think about only setting the values if it was different than the default

Something like
```rust
   let mut format = CsvFormat::default();
   if let Some(quote) = statement_options.take_str_option("quote") {
    format = format.with_quote(quote)
   }
   ...
```

?

**Review by alamb — APPROVED** [review]
Thank you @Asura7969  -- this looks great. I had a suggestion related to settings and it would be nice to add a reference to the comments, but I also think we could merge this PR as is and do them as a follow on.

Let me know what you would like to do.

Thanks again!

File: datafusion/core/src/datasource/listing_table_factory.rs (hunk 3 above)
Comment (Asura7969) [reply datafusion/core/src/datasource/listing_table_factory.rs:74]: you are right

**Comment by Asura7969** [comment]
> Thank you @Asura7969 -- this looks great. I had a suggestion related to settings and it would be nice to add a reference to the comments, but I also think we could merge this PR as is and do them as a follow on.
> 
> Let me know what you would like to do.
> 
> Thanks again!

I will follow your suggestions

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thank you @Asura7969 !

**Pull request merged**